### ISSUE TITLE
Modifying YOLOv8 for 6D Pose Estimation: Where Should I Start?

### ISSUE THREAD

**Issue by piyanuoyubuliao** [ultralytics/ultralytics#20084]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I want to modify the existing code to handle a new task (6D pose estimation). I currently know that I need to modify the dataloader, augmentation, head, and loss modules. However, the codebase is highly integrated, and I don't know where to start. Could you provide me with some guidance or relevant documentation? Thanks！

**Comment by UltralyticsAssistant** [comment]
👋 Hello @piyanuoyubuliao, thank you for your interest in Ultralytics 🚀! We appreciate your question about modifying YOLO for 6D pose estimation.

We recommend exploring the [Docs](https://docs.ultralytics.com/) for an in-depth understanding of the YOLO architecture and codebase. You can also find useful examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage there. These resources may help you get started with the modifications you're planning.

If this is a 🐛 Bug Report, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to assist in debugging. For custom modifications ❓ like yours, please include as much detail as possible, such as specific modules you're working on, relevant code snippets, or logs, so our team can better understand your requirements.

Additionally, feel free to connect with the Ultralytics community for further discussion and support:
- Join our [Discord](https://discord.com/invite/ultralytics) server for real-time conversation 🎧.
- Participate in long-form discussions on [Discourse](https://community.ultralytics.com/).
- Share insights or questions on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please ensure you're using the latest version of the `ultralytics` package. Update it along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to avoid potential issues:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments, all pre-configured with dependencies:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. These tests ensure reliable operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across platforms.

This is an automated response to guide you in the right direction. An Ultralytics engineer will review your query and provide assistance soon. Stay tuned! 😊

**Comment by glenn-jocher** [comment]
Hi @piyanuoyubuliao,

Modifying YOLOv8 for 6D pose estimation is certainly possible but requires careful changes in several components. Here's where to start:

1. Study the existing pose estimation implementation in `ultralytics/models/yolo/pose/` directory to understand the current architecture
2. For the dataloader modifications, check `ultralytics/data/dataset.py` and `ultralytics/data/build.py`
3. Augmentation code is in `ultralytics/data/augment.py`
4. Loss functions for pose are in `ultralytics/utils/loss.py` (see `v8PoseLoss` class)
5. Model head modifications would be in `ultralytics/nn/tasks.py` (see `PoseModel` class)

I recommend first creating a custom task that extends the existing pose estimation task, then gradually modifying it to support 6D outputs. You'll need to adapt the output format, loss functions, and evaluation metrics for 6D pose.

If you have more specific implementation questions as you progress, feel free to ask.

**Comment by piyanuoyubuliao** [comment]
**How to pass camera intrinsics and CAD model points in Ultralytics 6D pose pipeline?**

Hi Ultralytics team,
I am working on a 6D pose estimation task (estimating object rotation R and translation T) using Ultralytics models, and I have a question about the data flow design.
For 6D pose estimation, training and augmentation require additional geometric inputs:
**- Camera intrinsics matrix K
- CAD model point cloud / 3D model points**
These are needed for:
- **Loss computation** (e.g. reprojection / ADD(-S) style losses)
- **Image augmentations** such as letterbox, resize, crop, etc., which require updating K (or equivalently R/T) to keep pose labels consistent.
Currently, it is not clear how K and CAD points should be propagated through:
- Dataset → Dataloader → Model → Loss
- Image augmentation pipeline
**What is the recommended way to integrate camera intrinsics K and CAD model points into Ultralytics’ data flow?**
For example, should they be attached to the batch dict, Instances / extra fields, or handled in a custom geometry-aware transform?
Any guidance or best practices would be appreciated. Thanks for your great work!

**Comment by glenn-jocher** [comment]
Since this issue is closed/marked stale, please open a new GitHub issue (or a post in [Ultralytics Discussions](https://github.com/orgs/ultralytics/discussions)) with `yolo checks`, the full traceback (if any), and a minimal example of how you want to store `K`, `R|t`, and CAD-point IDs so it’s visible to the team; implementation-wise the cleanest pattern is to carry `K` (and a lightweight `cad_id`/`class_id`, not full point clouds) through the same `batch` dict that the pose pipeline already consumes (see how `batch["keypoints"]`/`batch["batch_idx"]` are used in the [PoseLoss26 / v8PoseLoss reference](https://docs.ultralytics.com/reference/utils/loss/)), then write a custom geometry-aware transform that returns the applied resize/pad/affine so you can update `K` (and any projected labels) consistently with the image.
